internal/v5/auth.go

```diff
@@ -296,15 +295,33 @@ func areAllowedEntities(entityIds []*router.ResolvedURL, allowedEntities string)
 // AuthorizeEntity checks that the given HTTP request
 // can access the entity with the given id.
 func (h *ReqHandler) AuthorizeEntity(id *router.ResolvedURL, req *http.Request) error {
-	baseEntity, err := h.Cache.BaseEntity(id.UserOwnedURL(), charmstore.FieldSelector("acls"))
+	acls, err := h.entityACL(id)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
+	}
+	return h.authorizeWithPerms(req, acls.Read, acls.Write, id)
+}
+
+func (h *ReqHandler) entityACL(id *router.ResolvedURL) (mongodoc.ACL, error) {
+	entity, err := h.Cache.Entity(id.UserOwnedURL(), charmstore.FieldSelector("development", "stable"))
 	if err != nil {
 		if errgo.Cause(err) == params.ErrNotFound {
-			return errgo.WithCausef(nil, params.ErrNotFound, "entity %q not found", id)
+			return mongodoc.ACL{}, errgo.WithCausef(nil, params.ErrNotFound, "entity %q not found", id)
 		}
-		return errgo.Notef(err, "cannot retrieve entity %q for authorization", id)
+		return mongodoc.ACL{}, errgo.Notef(err, "cannot retrieve entity %q for authorization", id)
+	}
+	baseEntity, err := h.Cache.BaseEntity(id.UserOwnedURL(), charmstore.FieldSelector("acls", "developmentacls", "stableacls"))
+	if err != nil {
+		return mongodoc.ACL{}, errgo.Notef(err, "cannot retrieve base entity %q for authorization", id)
 	}
 	acls := baseEntity.ACLs
-	return h.authorizeWithPerms(req, acls.Read, acls.Write, id)
+	if entity.Development {
```

Comment:
So why we are adding to the unpublished permissions instead of just switching the ACLs entirely?
For example, in the promulgation case, we might want to remove write permissions on the stable ACLs to the original owner but still allow him to upload new unpublished revisions. Given that we should start with the same permissions on the three ACLs, and that therefore initially the user has permissions on all channels, is this really required? If so, please let's add a comment about why we are doing this?